File: src/test/java/net/flyingfishflash/loremlist/unit/core/response/advice/ApiExceptionHandlerTests.java
package net.flyingfishflash.loremlist.unit.core.response.advice;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.validation.ConstraintViolationException;
import net.flyingfishflash.loremlist.core.exceptions.CoreException;
import net.flyingfishflash.loremlist.core.response.advice.CoreExceptionHandler;
import net.flyingfishflash.loremlist.core.response.structure.ResponseProblem;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.core.env.Environment;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.http.converter.HttpMessageNotReadableException;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.context.request.ServletWebRequest;
import org.springframework.web.method.annotation.HandlerMethodValidationException;

class ApiExceptionHandlerTests {

  private static Environment mockEnvironment(String includeStacktrace) {
    Environment mockEnvironment = mock(Environment.class);
    when(mockEnvironment.getProperty("server.error.include-stacktrace"))
        .thenReturn(includeStacktrace);
    return mockEnvironment;
  }

  private static ResponseProblem body(ResponseEntity<?> responseEntity) {
    assertThat(responseEntity.getBody()).isInstanceOf(ResponseProblem.class);
    return (ResponseProblem) responseEntity.getBody();
  }

  private static HttpServletRequest mockHttpServletRequest() {
    HttpServletRequest mockHttpServletRequest = mock(HttpServletRequest.class);
    when(mockHttpServletRequest.getMethod()).thenReturn("GET");
    when(mockHttpServletRequest.getRequestURI()).thenReturn("/irrelevant");
    return mockHttpServletRequest;
  }

  private static ServletWebRequest mockWebRequest() {
    ServletWebRequest mockWebRequest = mock(ServletWebRequest.class);
    HttpServletRequest request = mockHttpServletRequest();
    when(mockWebRequest.getRequest()).thenReturn(request);
    when(mockWebRequest.getDescription(false)).thenReturn("uri=/irrelevant");
    return mockWebRequest;
  }

  @Nested
  class HandleAbstractApiException {

    @Test
    void extensionsNone() {
      HttpServletRequest mockHttpServletRequest = mockHttpServletRequest();
      CoreException apiException = CoreException.builder().build();
      CoreExceptionHandler coreExceptionHandler =
          new CoreExceptionHandler(mockEnvironment("never"));
      ResponseEntity<ResponseProblem> responseEntity =
          coreExceptionHandler.handleAbstractCoreException(mockHttpServletRequest, apiException);
      assertThat(responseEntity.getBody().content().cause()).isNull();
      assertThat(responseEntity.getBody().content().stackTrace()).isNull();
      assertThat(responseEntity.getBody().content().validationErrors()).isNull();
    }

    @Test
    void extensionsCause() {
      HttpServletRequest mockHttpServletRequest = mockHttpServletRequest();
      CoreException apiException = CoreException.builder().cause(new RuntimeException()).build();
      CoreExceptionHandler coreExceptionHandler =
          new CoreExceptionHandler(mockEnvironment("never"));
      ResponseEntity<ResponseProblem> responseEntity =
          coreExceptionHandler.handleAbstractCoreException(mockHttpServletRequest, apiException);
      assertThat(responseEntity.getBody().content().cause()).isNotNull();
      assertThat(responseEntity.getBody().content().stackTrace()).isNull();
      assertThat(responseEntity.getBody().content().validationErrors()).isNull();
    }

    @Test
    void extensionsStacktrace() {
      HttpServletRequest mockHttpServletRequest = mockHttpServletRequest();
      CoreException apiException = CoreException.builder().build();
      CoreExceptionHandler coreExceptionHandler =
          new CoreExceptionHandler(mockEnvironment("always"));
      ResponseEntity<ResponseProblem> responseEntity =
          coreExceptionHandler.handleAbstractCoreException(mockHttpServletRequest, apiException);
      assertThat(responseEntity.getBody().content().cause()).isNull();
      assertThat(responseEntity.getBody().content().stackTrace()).isNotNull();
      assertThat(responseEntity.getBody().content().validationErrors()).isNull();
    }

    @Test
    void extensionsStacktraceCauseWithNoMessage() {
      HttpServletRequest mockHttpServletRequest = mockHttpServletRequest();
      CoreException apiException = CoreException.builder().cause(new RuntimeException()).build();
      CoreExceptionHandler coreExceptionHandler =
          new CoreExceptionHandler(mockEnvironment("always"));
      ResponseEntity<ResponseProblem> responseEntity =
          coreExceptionHandler.handleAbstractCoreException(mockHttpServletRequest, apiException);
      assertThat(responseEntity.getBody().content().cause()).isNotNull();
      assertThat(responseEntity.getBody().content().stackTrace()).isNotNull();
      assertThat(responseEntity.getBody().content().validationErrors()).isNull();
      assertThat(responseEntity.getBody().content().cause().name()).isEqualTo("RuntimeException");
      assertThat(responseEntity.getBody().content().cause().message())
          .isEqualTo("Exception message not present.");
    }

    @Test
    void extensionsStacktraceCauseIsNestedException() {
      HttpServletRequest mockHttpServletRequest = mockHttpServletRequest();
      IllegalArgumentException rootCause = new IllegalArgumentException("Root cause exception");
      IllegalStateException intermediateCause =
          new IllegalStateException("Intermediate cause exception");
      intermediateCause.initCause(rootCause);
      CoreException topLevelException =
          CoreException.builder().message("Top level exception").cause(intermediateCause).build();
      CoreExceptionHandler coreExceptionHandler =
          new CoreExceptionHandler(mockEnvironment("always"));
      ResponseEntity<ResponseProblem> responseEntity =
          coreExceptionHandler.handleAbstractCoreException(
              mockHttpServletRequest, topLevelException);
      // Note: CoreException.supplemental is only ever populated via an explicit
      // builder().supplemental(...) call; this exception never sets it, so it stays null here.
      assertThat(responseEntity.getBody().content().cause().name())
          .isEqualTo("IllegalArgumentException");
      assertThat(responseEntity.getBody().content().cause().message())
          .isEqualTo("Root cause exception");
      assertThat(responseEntity.getBody().content().stackTrace()).isNotNull();
      assertThat(responseEntity.getBody().content().validationErrors()).isNull();
    }
  }

  @Nested
  class HandleException {

    @Test
    void withStacktrace() {
      HttpServletRequest mockHttpServletRequest = mockHttpServletRequest();
      Exception exception = new Exception();
      CoreExceptionHandler coreExceptionHandler =
          new CoreExceptionHandler(mockEnvironment("always"));
      ResponseEntity<ResponseProblem> responseEntity =
          coreExceptionHandler.handleException(mockHttpServletRequest, exception);
      assertThat(responseEntity.getBody().content().validationErrors()).isNull();
      assertThat(responseEntity.getBody().content().stackTrace()).isNotNull();
    }

    @Test
    void withoutStacktrace() {
      HttpServletRequest mockHttpServletRequest = mockHttpServletRequest();
      Exception exception = new Exception();
      CoreExceptionHandler coreExceptionHandler = new CoreExceptionHandler(mockEnvironment(null));
      ResponseEntity<ResponseProblem> responseEntity =
          coreExceptionHandler.handleException(mockHttpServletRequest, exception);
      assertThat(responseEntity.getBody().content().validationErrors()).isNull();
      assertThat(responseEntity.getBody().content().stackTrace()).isNull();
    }

    @Test
    void withDetailMessage() {
      HttpServletRequest mockHttpServletRequest = mockHttpServletRequest();
      Exception exception = new Exception("Lorem Ipsum");
      CoreExceptionHandler coreExceptionHandler = new CoreExceptionHandler(mockEnvironment(null));
      ResponseEntity<ResponseProblem> responseEntity =
          coreExceptionHandler.handleException(mockHttpServletRequest, exception);
      assertThat(responseEntity.getBody().content().detail()).isEqualTo("Lorem Ipsum");
    }
  }

  @Nested
  class HandleHttpMessageNotReadable {

    @Test
    void withStacktrace() throws Exception {
      HttpMessageNotReadableException mockException = mock(HttpMessageNotReadableException.class);
      when(mockException.getStackTrace()).thenReturn(new StackTraceElement[0]);
      HttpHeaders mockHttpHeaders = mock(HttpHeaders.class);
      ServletWebRequest mockWebRequest = mockWebRequest();
      CoreExceptionHandler coreExceptionHandler =
          new CoreExceptionHandler(mockEnvironment("always"));
      ResponseEntity<Object> responseEntity =
          coreExceptionHandler.handleHttpMessageNotReadable(
              mockException, mockHttpHeaders, HttpStatus.I_AM_A_TEAPOT, mockWebRequest);
      assertThat(body(responseEntity).content().stackTrace()).isNotNull();
      assertThat(body(responseEntity).content().validationErrors()).isNull();
    }

    @Test
    void withoutStacktrace() throws Exception {
      HttpMessageNotReadableException mockException = mock(HttpMessageNotReadableException.class);
      HttpHeaders mockHttpHeaders = mock(HttpHeaders.class);
      ServletWebRequest mockWebRequest = mockWebRequest();
      CoreExceptionHandler coreExceptionHandler = new CoreExceptionHandler(mockEnvironment(null));
      ResponseEntity<Object> responseEntity =
          coreExceptionHandler.handleHttpMessageNotReadable(
              mockException, mockHttpHeaders, HttpStatus.I_AM_A_TEAPOT, mockWebRequest);
      assertThat(body(responseEntity).content().stackTrace()).isNull();
      assertThat(body(responseEntity).content().validationErrors()).isNull();
    }

    @Test
    void withoutDetailMessage() throws Exception {
      HttpMessageNotReadableException mockException = mock(HttpMessageNotReadableException.class);
      when(mockException.getMessage()).thenReturn(null);
      HttpHeaders mockHttpHeaders = mock(HttpHeaders.class);
      ServletWebRequest mockWebRequest = mockWebRequest();
      CoreExceptionHandler coreExceptionHandler = new CoreExceptionHandler(mockEnvironment(null));
      ResponseEntity<Object> responseEntity =
          coreExceptionHandler.handleHttpMessageNotReadable(
              mockException, mockHttpHeaders, HttpStatus.I_AM_A_TEAPOT, mockWebRequest);
      assertThat(body(responseEntity).content().detail())
          .isEqualTo(CoreExceptionHandler.EXCEPTION_MESSAGE_NOT_PRESENT);
      assertThat(body(responseEntity).content().stackTrace()).isNull();
      assertThat(body(responseEntity).content().validationErrors()).isNull();
    }
  }

  @Nested
  class HandleHandlerMethodValidationException {

    @Test
    void withStacktrace() {
      HandlerMethodValidationException mockException = mock(HandlerMethodValidationException.class);
      when(mockException.getStackTrace()).thenReturn(new StackTraceElement[0]);
      HttpHeaders mockHttpHeaders = mock(HttpHeaders.class);
      ServletWebRequest mockWebRequest = mockWebRequest();
      CoreExceptionHandler coreExceptionHandler =
          new CoreExceptionHandler(mockEnvironment("always"));
      ResponseEntity<Object> responseEntity =
          coreExceptionHandler.handleHandlerMethodValidationException(
              mockException, mockHttpHeaders, HttpStatus.I_AM_A_TEAPOT, mockWebRequest);
      assertThat(body(responseEntity).content().stackTrace()).isNotNull();
      assertThat(body(responseEntity).content().validationErrors()).isNotNull();
    }

    @Test
    void withoutStacktrace() {
      HandlerMethodValidationException mockException = mock(HandlerMethodValidationException.class);
      HttpHeaders mockHttpHeaders = mock(HttpHeaders.class);
      ServletWebRequest mockWebRequest = mockWebRequest();
      CoreExceptionHandler coreExceptionHandler = new CoreExceptionHandler(mockEnvironment(null));
      ResponseEntity<Object> responseEntity =
          coreExceptionHandler.handleHandlerMethodValidationException(
              mockException, mockHttpHeaders, HttpStatus.I_AM_A_TEAPOT, mockWebRequest);
      assertThat(body(responseEntity).content().stackTrace()).isNull();
      assertThat(body(responseEntity).content().validationErrors()).isNotNull();
    }
  }

  @Nested
  class HandleConstraintViolationException {

    @Test
    void withStacktrace() {
      ConstraintViolationException mockException = mock(ConstraintViolationException.class);
      when(mockException.getCause()).thenReturn(new Exception("Constraint Violation Cause"));
      when(mockException.getStackTrace()).thenReturn(new StackTraceElement[0]);
      HttpServletRequest mockHttpServletRequest = mockHttpServletRequest();
      CoreExceptionHandler coreExceptionHandler =
          new CoreExceptionHandler(mockEnvironment("always"));
      ResponseEntity<ResponseProblem> responseEntity =
          coreExceptionHandler.handleConstraintViolationException(
              mockHttpServletRequest, mockException);
      assertThat(responseEntity.getBody().content().stackTrace()).isNotNull();
      assertThat(responseEntity.getBody().content().validationErrors()).isNotNull();
    }

    @Test
    void withoutStacktrace() {
      ConstraintViolationException mockException = mock(ConstraintViolationException.class);
      when(mockException.getCause()).thenReturn(new Exception("Constraint Violation Cause"));
      HttpServletRequest mockHttpServletRequest = mockHttpServletRequest();
      CoreExceptionHandler coreExceptionHandler = new CoreExceptionHandler(mockEnvironment(null));
      ResponseEntity<ResponseProblem> responseEntity =
          coreExceptionHandler.handleConstraintViolationException(
              mockHttpServletRequest, mockException);
      assertThat(responseEntity.getBody().content().stackTrace()).isNull();
      assertThat(responseEntity.getBody().content().validationErrors()).isNotNull();
    }
  }

  @Nested
  class HandleMethodArgumentNotValid {

    @Test
    void withStacktrace() {
      MethodArgumentNotValidException mockException = mock(MethodArgumentNotValidException.class);
      when(mockException.getStackTrace()).thenReturn(new StackTraceElement[0]);
      org.springframework.validation.BindingResult mockBindingResult =
          mock(org.springframework.validation.BindingResult.class);
      when(mockBindingResult.getFieldErrors()).thenReturn(java.util.List.of());
      when(mockException.getBindingResult()).thenReturn(mockBindingResult);
      HttpHeaders mockHttpHeaders = mock(HttpHeaders.class);
      ServletWebRequest mockWebRequest = mockWebRequest();
      CoreExceptionHandler coreExceptionHandler =
          new CoreExceptionHandler(mockEnvironment("always"));
      ResponseEntity<Object> responseEntity =
          coreExceptionHandler.handleMethodArgumentNotValid(
              mockException, mockHttpHeaders, HttpStatus.I_AM_A_TEAPOT, mockWebRequest);
      assertThat(body(responseEntity).content().stackTrace()).isNotNull();
      assertThat(body(responseEntity).content().validationErrors()).isNotNull();
    }

    @Test
    void withoutStacktrace() {
      MethodArgumentNotValidException mockException = mock(MethodArgumentNotValidException.class);
      org.springframework.validation.BindingResult mockBindingResult =
          mock(org.springframework.validation.BindingResult.class);
      when(mockBindingResult.getFieldErrors()).thenReturn(java.util.List.of());
      when(mockException.getBindingResult()).thenReturn(mockBindingResult);
      HttpHeaders mockHttpHeaders = mock(HttpHeaders.class);
      ServletWebRequest mockWebRequest = mockWebRequest();
      CoreExceptionHandler coreExceptionHandler = new CoreExceptionHandler(mockEnvironment(null));
      ResponseEntity<Object> responseEntity =
          coreExceptionHandler.handleMethodArgumentNotValid(
              mockException, mockHttpHeaders, HttpStatus.I_AM_A_TEAPOT, mockWebRequest);
      assertThat(body(responseEntity).content().stackTrace()).isNull();
      assertThat(body(responseEntity).content().validationErrors()).isNotNull();
    }
  }
}

File: src/test/kotlin/net/flyingfishflash/loremlist/unit/core/response/advice/ApiExceptionHandlerTests.kt
package net.flyingfishflash.loremlist.unit.core.response.advice

import io.kotest.core.spec.style.DescribeSpec
import io.kotest.matchers.nulls.shouldBeNull
import io.kotest.matchers.nulls.shouldNotBeNull
import io.kotest.matchers.shouldBe
import io.kotest.matchers.types.shouldBeInstanceOf
import io.mockk.clearAllMocks
import io.mockk.every
import io.mockk.mockk
import io.mockk.unmockkAll
import jakarta.servlet.http.HttpServletRequest
import jakarta.validation.ConstraintViolationException
import net.flyingfishflash.loremlist.core.exceptions.ApiException
import net.flyingfishflash.loremlist.core.response.advice.ApiExceptionHandler
import net.flyingfishflash.loremlist.core.response.structure.ResponseProblem
import org.springframework.core.env.Environment
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.converter.HttpMessageNotReadableException
import org.springframework.web.bind.MethodArgumentNotValidException
import org.springframework.web.context.request.ServletWebRequest
import org.springframework.web.method.annotation.HandlerMethodValidationException

class ApiExceptionHandlerTests : DescribeSpec({

  afterEach { clearAllMocks() }
  afterSpec { unmockkAll() }

  describe("handleAbstractApiException()") {

    it("extensions: none") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns "never"
      val mockHttpServletRequest = mockk<HttpServletRequest>(relaxed = true)
      val apiException = ApiException()
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleAbstractApiException(mockHttpServletRequest, apiException)
      responseEntity.body?.content?.cause.shouldBeNull()
      responseEntity.body?.content?.stackTrace.shouldBeNull()
      responseEntity.body?.content?.validationErrors.shouldBeNull()
    }

    it("extensions: cause") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns "never"
      val mockHttpServletRequest = mockk<HttpServletRequest>(relaxed = true)
      val apiException = ApiException(cause = RuntimeException())
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleAbstractApiException(mockHttpServletRequest, apiException)
      responseEntity.body?.content?.cause.shouldNotBeNull()
      responseEntity.body?.content?.stackTrace.shouldBeNull()
      responseEntity.body?.content?.validationErrors.shouldBeNull()
    }

    it("extensions: stacktrace") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns "always"
      val mockHttpServletRequest = mockk<HttpServletRequest>(relaxed = true)
      val apiException = ApiException()
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleAbstractApiException(mockHttpServletRequest, apiException)
      responseEntity.body?.content?.cause.shouldBeNull()
      responseEntity.body?.content?.stackTrace.shouldNotBeNull()
      responseEntity.body?.content?.validationErrors.shouldBeNull()
    }

    it("extensions: stacktrace, cause wth no message") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns "always"
      val mockHttpServletRequest = mockk<HttpServletRequest>(relaxed = true)
      val apiException = ApiException(cause = RuntimeException())
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleAbstractApiException(mockHttpServletRequest, apiException)
      responseEntity.body?.content?.cause.shouldNotBeNull()
      responseEntity.body?.content?.stackTrace.shouldNotBeNull()
      responseEntity.body?.content?.validationErrors.shouldBeNull()
      responseEntity.body?.content?.cause?.name.shouldBe("RuntimeException")
      responseEntity.body?.content?.cause?.message.shouldBe("Exception message not present.")
    }

    it("extensions: stacktrace, cause is nested exception") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns "always"
      val mockHttpServletRequest = mockk<HttpServletRequest>(relaxed = true)
      val rootCause = IllegalArgumentException("Root cause exception")
      val intermediateCause = IllegalStateException("Intermediate cause exception")
      intermediateCause.initCause(rootCause)
      val topLevelException = ApiException(message = "Top level exception", cause = intermediateCause)
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleAbstractApiException(mockHttpServletRequest, topLevelException)
      responseEntity.body?.content?.supplemental?.shouldNotBeNull()
      responseEntity.body?.content?.cause?.name.shouldBe("IllegalArgumentException")
      responseEntity.body?.content?.cause?.message.shouldBe("Root cause exception")
      responseEntity.body?.content?.stackTrace.shouldNotBeNull()
      responseEntity.body?.content?.validationErrors.shouldBeNull()
    }
  }

  describe("handleException()") {
    it("with stacktrace") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns "always"
      val mockHttpServletRequest = mockk<HttpServletRequest>(relaxed = true)
      val exception = Exception()
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleException(mockHttpServletRequest, exception)
      responseEntity.body?.content?.validationErrors.shouldBeNull()
      responseEntity.body?.content?.stackTrace.shouldNotBeNull()
    }

    it("without stacktrace") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns null
      val mockHttpServletRequest = mockk<HttpServletRequest>(relaxed = true)
      val exception = Exception()
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleException(mockHttpServletRequest, exception)
      responseEntity.body?.content?.validationErrors.shouldBeNull()
      responseEntity.body?.content?.stackTrace.shouldBeNull()
    }

    it("with detail message") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns null
      val mockHttpServletRequest = mockk<HttpServletRequest>(relaxed = true)
      val exception = Exception("Lorem Ipsum")
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleException(mockHttpServletRequest, exception)
      responseEntity.body?.content?.detail.shouldBe("Lorem Ipsum")
    }
  }

  describe("handleHttpMessageNotReadable()") {
    it("with stacktrace") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns "always"
      val mockHttpMessageNotReadableException = mockk<HttpMessageNotReadableException>(relaxed = true)
      val mockHttpHeaders = mockk<HttpHeaders>(relaxed = true)
      val mockWebRequest = mockk<ServletWebRequest>(relaxed = true)
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleHttpMessageNotReadable(
        mockHttpMessageNotReadableException,
        mockHttpHeaders,
        HttpStatus.I_AM_A_TEAPOT,
        mockWebRequest,
      )
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.stackTrace.shouldNotBeNull()
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.validationErrors.shouldBeNull()
    }

    it("without stacktrace") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns null
      val mockHttpMessageNotReadableException = mockk<HttpMessageNotReadableException>(relaxed = true)
      val mockHttpHeaders = mockk<HttpHeaders>(relaxed = true)
      val mockWebRequest = mockk<ServletWebRequest>(relaxed = true)
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleHttpMessageNotReadable(
        mockHttpMessageNotReadableException,
        mockHttpHeaders,
        HttpStatus.I_AM_A_TEAPOT,
        mockWebRequest,
      )
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.stackTrace.shouldBeNull()
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.validationErrors.shouldBeNull()
    }

    it("without detail message") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns null
      val mockHttpMessageNotReadableException = mockk<HttpMessageNotReadableException>(relaxed = true)
      every { mockHttpMessageNotReadableException.message } returns null
      val mockHttpHeaders = mockk<HttpHeaders>(relaxed = true)
      val mockWebRequest = mockk<ServletWebRequest>(relaxed = true)
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleHttpMessageNotReadable(
        mockHttpMessageNotReadableException,
        mockHttpHeaders,
        HttpStatus.I_AM_A_TEAPOT,
        mockWebRequest,
      )
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.detail.shouldBe(
        ApiExceptionHandler.EXCEPTION_MESSAGE_NOT_PRESENT,
      )
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.stackTrace.shouldBeNull()
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.validationErrors.shouldBeNull()
    }
  }

  describe("handleHandlerMethodValidationException()") {
    it("with stacktrace") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns "always"
      val mockHandlerMethodValidationException = mockk<HandlerMethodValidationException>(relaxed = true)
      val mockHttpHeaders = mockk<HttpHeaders>(relaxed = true)
      val mockWebRequest = mockk<ServletWebRequest>(relaxed = true)
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleHandlerMethodValidationException(
        mockHandlerMethodValidationException,
        mockHttpHeaders,
        HttpStatus.I_AM_A_TEAPOT,
        mockWebRequest,
      )
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.stackTrace.shouldNotBeNull()
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.validationErrors.shouldNotBeNull()
    }

    it("without stacktrace") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns null
      val mockHandlerMethodValidationException = mockk<HandlerMethodValidationException>(relaxed = true)
      val mockHttpHeaders = mockk<HttpHeaders>(relaxed = true)
      val mockWebRequest = mockk<ServletWebRequest>(relaxed = true)
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleHandlerMethodValidationException(
        mockHandlerMethodValidationException,
        mockHttpHeaders,
        HttpStatus.I_AM_A_TEAPOT,
        mockWebRequest,
      )
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.stackTrace.shouldBeNull()
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.validationErrors.shouldNotBeNull()
    }
  }

  describe("handleConstraintViolationException()") {
    it("with stacktrace") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns "always"
      val mockConstraintViolationException = mockk<ConstraintViolationException>(relaxed = true)
      every { mockConstraintViolationException.cause } returns Exception("Constraint Violation Cause")
      val mockHttpServletRequest = mockk<HttpServletRequest>(relaxed = true)
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleConstraintViolationException(
        mockHttpServletRequest,
        mockConstraintViolationException,
      )
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.stackTrace.shouldNotBeNull()
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.validationErrors.shouldNotBeNull()
    }

    it("without stacktrace") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns null
      val mockConstraintViolationException = mockk<ConstraintViolationException>(relaxed = true)
      every { mockConstraintViolationException.cause } returns Exception("Constraint Violation Cause")
      val mockHttpServletRequest = mockk<HttpServletRequest>(relaxed = true)
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleConstraintViolationException(
        mockHttpServletRequest,
        mockConstraintViolationException,
      )
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.stackTrace.shouldBeNull()
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.validationErrors.shouldNotBeNull()
    }
  }

  describe("handleMethodArgumentNotValid()") {
    it("with stacktrace") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns "always"
      val mockMethodArgumentNotValidException = mockk<MethodArgumentNotValidException>(relaxed = true)
      val mockHttpHeaders = mockk<HttpHeaders>(relaxed = true)
      val mockWebRequest = mockk<ServletWebRequest>(relaxed = true)
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleMethodArgumentNotValid(
        mockMethodArgumentNotValidException,
        mockHttpHeaders,
        HttpStatus.I_AM_A_TEAPOT,
        mockWebRequest,
      )
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.stackTrace.shouldNotBeNull()
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.validationErrors.shouldNotBeNull()
    }

    it("without stacktrace") {
      val mockEnvironment = mockk<Environment>(relaxed = true)
      every { mockEnvironment.getProperty("server.error.include-stacktrace") } returns null
      val mockMethodArgumentNotValidException = mockk<MethodArgumentNotValidException>(relaxed = true)
      val mockHttpHeaders = mockk<HttpHeaders>(relaxed = true)
      val mockWebRequest = mockk<ServletWebRequest>(relaxed = true)
      val apiExceptionHandler = ApiExceptionHandler(mockEnvironment)
      val responseEntity = apiExceptionHandler.handleMethodArgumentNotValid(
        mockMethodArgumentNotValidException,
        mockHttpHeaders,
        HttpStatus.I_AM_A_TEAPOT,
        mockWebRequest,
      )
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.stackTrace.shouldBeNull()
      responseEntity.body?.shouldNotBeNull().shouldBeInstanceOf<ResponseProblem>().content.validationErrors.shouldNotBeNull()
    }
  }
})
